pkg/distributor/query.go

```diff
@@ -428,6 +389,79 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSets [
 	return resp, nil
 }
 
+func (r *ingesterQueryResult) receiveResponse(stream ingester_client.Ingester_QueryStreamClient, queryLimiter *limiter.QueryLimiter) ([]labels.Labels, bool, error) {
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, false, err
+	}
+	defer resp.FreeBuffer()
+
+	if len(resp.Timeseries) > 0 {
+		for _, series := range resp.Timeseries {
+			if limitErr := queryLimiter.AddSeries(mimirpb.FromLabelAdaptersToLabels(series.Labels)); limitErr != nil {
+				return nil, false, limitErr
+			}
+		}
+
+		for i := range resp.Timeseries {
+			resp.Timeseries[i].CloneRefs()
+		}
+		r.timeseriesBatches = append(r.timeseriesBatches, resp.Timeseries)
+	} else if len(resp.Chunkseries) > 0 {
+		// Enforce the max chunks limits.
+		if err := queryLimiter.AddChunks(ingester_client.ChunksCount(resp.Chunkseries)); err != nil {
+			return nil, false, err
+		}
+
+		if err := queryLimiter.AddEstimatedChunks(ingester_client.ChunksCount(resp.Chunkseries)); err != nil {
+			return nil, false, err
+		}
+
+		for _, series := range resp.Chunkseries {
+			if err := queryLimiter.AddSeries(mimirpb.FromLabelAdaptersToLabels(series.Labels)); err != nil {
+				return nil, false, err
+			}
+		}
+
+		if err := queryLimiter.AddChunkBytes(ingester_client.ChunksSize(resp.Chunkseries)); err != nil {
+			return nil, false, err
+		}
+
+		for i := range resp.Chunkseries {
+			resp.Chunkseries[i].CloneRefs()
+		}
+		r.chunkseriesBatches = append(r.chunkseriesBatches, resp.Chunkseries)
+	} else if len(resp.StreamingSeries) > 0 {
+		labelsBatch := make([]labels.Labels, 0, len(resp.StreamingSeries))
+		var (
+			labelsBuilder labels.ScratchBuilder
+			l             labels.Labels
+		)
+		for _, s := range resp.StreamingSeries {
+			l = mimirpb.FromLabelAdaptersOverwriteLabelsSafe(&labelsBuilder, s.Labels, &l)
```

Comment:
The reuse and overwriting here makes me nervous and is a little difficult to reason about. Given we were previously using `mimirpb.FromLabelAdaptersToLabels(s.Labels)` here (which does a copy of the label names and values with `stringlabels` anyway, and so is equivalent to `FromLabelAdaptersToLabelsWithCopy`), what if we just use `FromLabelAdaptersToLabelsWithCopy` here?

Reply:
As discussed offline, I can extract this potential optimization and raise it separately from this PR.

Reply:
Done.